rust/src/utils.rs

```diff
@@ -304,6 +310,101 @@ pub(crate) fn translate_path_for_ostree_impl(path: &str) -> Option<String> {
     None
 }
 
+#[context("Deduplicate tmpfiles entries")]
+pub fn deduplicate_tmpfiles_entries(tmprootfs_dfd: i32) -> CxxResult<()> {
+    let tmprootfs_dfd = unsafe { ffiutil::ffi_dirfd(tmprootfs_dfd)? };
+    static TMPFILESD: &str = "usr/lib/tmpfiles.d";
+    static RPMOSTREE_TMPFILESD: &str = "usr/lib/rpm-ostree/tmpfiles.d";
+    static AUTOVAR_PATH: &str = "rpm-ostree-autovar.conf";
+
+    let rpmostree_tmpfiles_list =
+        get_tmpfiles_path_list(&tmprootfs_dfd, RPMOSTREE_TMPFILESD).unwrap();
+    let system_tmpfiles_list = get_tmpfiles_path_list(&tmprootfs_dfd, TMPFILESD).unwrap();
+    if rpmostree_tmpfiles_list.is_empty() || system_tmpfiles_list.is_empty() {
+        println!("Not found any auto-generated or system tmpfiles.d config");
+        return Ok(());
+    }
+
+    // save rpm-ostree auto generated entries to hashmaps
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(RPMOSTREE_TMPFILESD)
+        .context("readdir {RPMOSTREE_TMPFILESD}")?;
+    let mut rpmostree_tmpfiles_entries =
+        save_tmpfile_entries(&tmpfiles_dir, &rpmostree_tmpfiles_list).unwrap();
+    // save system entries to hashmaps
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(TMPFILESD)
+        .context("readdir {TMPFILESD}")?;
+    let system_tmpfiles_entries =
+        save_tmpfile_entries(&tmpfiles_dir, &system_tmpfiles_list).unwrap();
+
+    // remove duplicated entries in auto-generated tmpfiles.d,
+    // which are already in system tmpfiles
+    for key in system_tmpfiles_entries.keys() {
+        if rpmostree_tmpfiles_entries.contains_key(key) {
+            rpmostree_tmpfiles_entries.remove(key);
+        }
+    }
+
+    {
+        // save the noduplicated entries
+        let mut entries = String::from("");
+        for (_key, value) in rpmostree_tmpfiles_entries {
+            entries.push_str(&format!("{}\n", value));
+        }
+
+        if tmpfiles_dir.try_exists(AUTOVAR_PATH)? {
+            tmpfiles_dir.remove_file(AUTOVAR_PATH)?
+        }
+        let perms = Permissions::from_mode(0o644);
+        tmpfiles_dir.atomic_write_with_perms(&AUTOVAR_PATH, entries.as_bytes(), perms)?;
+    }
+    Ok(())
+}
+
+#[context("Get tmpfiles files list")]
+pub(crate) fn get_tmpfiles_path_list(tmprootfs_dfd: &Dir, path: &str) -> Result<Vec<String>> {
+    let tmpfiles_dir = tmprootfs_dfd.open_dir(path).context("readdir {path}")?;
+    let mut tmpfiles_list = Vec::new();
+    for entry in tmpfiles_dir.entries()? {
+        let entry = entry?;
+        let name = entry.file_name();
+        let name = name.to_str().unwrap();
+        // skip README
+        if name == "README" || name == "rpm-ostree-autovar.conf" {
```

Comment:
Re. README: I think anyway we should only pick up `.conf` files.

Re. autovar: I think it'd be cleaner if we just upfront in `deduplicate_tmpfiles_entries()` always delete the file?